{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "137cca62-7604-4e4e-a695-41fdc9e19f7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline \n",
    "import numpy as np\n",
    "from matplotlib import pyplot as plt\n",
    "from scipy import stats\n",
    "import pandas as pd\n",
    "\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3bf815ed-2980-457e-91bc-93b8b8d6d394",
   "metadata": {},
   "source": [
    "### Define functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "821f22c8-0b45-480c-934b-04d184598f79",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def load_fasta(fasta_file,sites_of_interest):\n",
    "    \"\"\"\n",
    "    Load a fasta file into a pandas dataframe.\n",
    "    \n",
    "    fasta_file: fasta file (aligned, single line per sequence)\n",
    "    sites_of_interest: dictionary where indexes are position in alignment\n",
    "                       sequence (starting at 0) and values are sequence number\n",
    "                       in amino acid sequence.\n",
    "    \n",
    "    returns df\n",
    "    \"\"\"\n",
    "\n",
    "    # Read fasta file into a numpy array\n",
    "    names = []\n",
    "    sequences = []\n",
    "    with open(fasta_file) as f:\n",
    "        for line in f:\n",
    "            if line.startswith(\">\"):\n",
    "                names.append(line[1:].strip())\n",
    "                sequences.append([])\n",
    "                continue\n",
    "\n",
    "            sequences[-1].extend([c for c in line.strip()])\n",
    "\n",
    "    # Get sequences as string\n",
    "    sequence_as_str = [\"\".join(s) for s in sequences]\n",
    "            \n",
    "    sequences = np.array(sequences)\n",
    "\n",
    "    # Create array of sequences containing only columns for sites of interest\n",
    "    slicer = np.array(list(sites_of_interest.keys()),dtype=int)\n",
    "\n",
    "\n",
    "    sequences = sequences[:,slicer]\n",
    "    \n",
    "    # Dict to use as source for data frame\n",
    "    to_pd = {}\n",
    "    to_pd[\"name\"] = names\n",
    "    to_pd[\"paralog\"] = [n.split(\"|\")[0] for n in names]\n",
    "\n",
    "    # Load columns for sites of interest into dict\n",
    "    for i in range(len(slicer)):\n",
    "        to_pd[f\"site_{sites_of_interest[slicer[i]]}\"] = [s for s in sequences[:,i]]\n",
    "        \n",
    "    to_pd[\"sequence\"] = sequence_as_str\n",
    "\n",
    "    return pd.DataFrame(to_pd)\n",
    "\n",
    "def classify_and_color(df):\n",
    "    \"\"\"\n",
    "    Classify amino acids in dataframe and update dataframe. \n",
    "    \n",
    "    Returns df, aa_to_class dictionary and class_to_color dictionary\n",
    "    \"\"\"\n",
    "\n",
    "    # Create dictionary mapping amino acid name to amino acid class\n",
    "    membership = [([\"K\",\"R\",\"H\"],\"basic\",\"pink\"),\n",
    "                  ([\"E\",\"D\"],\"acidic\",\"firebrick\"),\n",
    "                  ([\"F\",\"W\",\"Y\"],\"aromatic\",\"yellow\"),\n",
    "                  ([\"S\",\"T\",\"N\",\"Q\",\"C\"],\"polar\",\"pink\"),\n",
    "                  ([\"A\",\"I\",\"L\",\"V\",\"M\"],\"hphobic\",\"gray\"),\n",
    "                  ([\"P\"],\"pro\",\"pink\"),\n",
    "                  ([\"G\"],\"gly\",\"pink\"),\n",
    "                  ([\"-\",\"X\"],\"gap\",\"pink\")]\n",
    "\n",
    "    aa_to_class = {}\n",
    "    class_to_color = {}\n",
    "    for i in range(len(membership)):\n",
    "        for aa in membership[i][0]:\n",
    "            aa_to_class[aa] = membership[i][1]\n",
    "            class_to_color[membership[i][1]] = membership[i][2]\n",
    "\n",
    "    for col in list(df.columns):\n",
    "        if not col.startswith(\"site_\"):\n",
    "            continue\n",
    "\n",
    "        class_col = []\n",
    "        for aa in df.loc[:,col]:\n",
    "            class_col.append(aa_to_class[aa])\n",
    "        df[f\"class_{col}\"] = class_col\n",
    "\n",
    "    return df, aa_to_class, class_to_color\n",
    "\n",
    "\n",
    "def get_top_n_classes(count_dict,n=2):\n",
    "    \"\"\"\n",
    "    Grab the most common n classes from a sequence dataframe and return the\n",
    "    class values and their counts.\n",
    "    \"\"\"\n",
    "    \n",
    "    sorted_keys = [(count_dict[k],k) for k in count_dict]\n",
    "    sorted_keys.sort(reverse=True)\n",
    "    sorted_keys = dict([(v[1],i) for i, v in enumerate(sorted_keys[:n])])\n",
    "    \n",
    "    return sorted_keys\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe4fdba3-d3ae-481f-a755-6e29feaf89a8",
   "metadata": {},
   "source": [
    "### Do calculation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87cf9c33-1909-4607-a88f-e4047e29d49f",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Load the alignment into the dataframe\n",
    "df = load_fasta(fasta_file=\"alignment.fasta\",\n",
    "                sites_of_interest={174:37,204:63}) \n",
    "\n",
    "# Classify amino acids in dataframe\n",
    "df, aa_to_class, class_to_color = classify_and_color(df)\n",
    "\n",
    "# Make order of df pretty\n",
    "df = df.loc[:,[\"name\",\"paralog\",\"site_37\",\"site_63\",\"class_site_37\",\"class_site_63\",\"sequence\"]]\n",
    "df.to_csv(\"table_s6_site37-63-analysis.csv\")\n",
    "\n",
    "# We're interested in site 63\n",
    "focal_site = 63\n",
    "\n",
    "# Colors for output pie charts showing coincidence of amino acids \n",
    "coincidence_colors = {(\"hphobic\",\"aromatic\"):\"green\",\n",
    "                      (\"aromatic\",\"aromatic\"):\"orange\",\n",
    "                      (\"aromatic\",\"hphobic\"):\"gray\",\n",
    "                      (\"hphobic\",\"hphobic\"):\"purple\"}\n",
    "\n",
    "out = []\n",
    "\n",
    "# For every paralog in the alignment\n",
    "for para in np.unique(df.paralog):\n",
    "    \n",
    "    out.append(f\"{para}\\n\")\n",
    "    \n",
    "    # Grab rows corresponding to this paralog\n",
    "    rows = df.loc[df.paralog == para,:]\n",
    "    \n",
    "    # Get top two classes of amino acids observed at the focal site\n",
    "    obs, counts = np.unique(rows.loc[:,f\"class_site_{focal_site}\"],return_counts=True)\n",
    "    focal_class_dict = dict(zip(obs,counts))\n",
    "    focal_classes = get_top_n_classes(focal_class_dict,2)\n",
    "\n",
    "    # Create a figure object\n",
    "    fig, ax = plt.subplots(3,\n",
    "                           len([c for c in df.columns if c.startswith(\"site_\")]),\n",
    "                           figsize=(3,6))\n",
    "    fig.suptitle(para)\n",
    "    \n",
    "    # Go through columns\n",
    "    counter = 0\n",
    "    for col in df.columns:\n",
    "        \n",
    "        # Skip non site_ columns\n",
    "        if not col.startswith(\"site_\"):\n",
    "            continue\n",
    "\n",
    "        # Create pie chart with counts for all amino acids\n",
    "        obs, counts = np.unique(rows.loc[:,col],return_counts=True)\n",
    "        count_dict = dict(zip(obs,counts))\n",
    "        ax[0,counter].pie(count_dict.values(),labels=count_dict.keys())\n",
    "        \n",
    "        # Create pie chart with counts for all amino acid classes\n",
    "        obs, counts = np.unique(rows.loc[:,f\"class_{col}\"],return_counts=True)\n",
    "        class_dict = dict(zip(obs,counts))        \n",
    "        local_classes = get_top_n_classes(class_dict,n=2)\n",
    "        ax[1,counter].pie(class_dict.values(),labels=class_dict.keys(),\n",
    "                  colors=[class_to_color[c] for c in class_dict.keys()])\n",
    "        \n",
    "        # If this is not site 37, don't make a coincidence pie chart\n",
    "        if col != \"site_37\":\n",
    "            ax[2,counter].axis(\"off\")\n",
    "            counter +=1\n",
    "            continue\n",
    "        \n",
    "        # Get classes seen for focal (63) and local (37) site\n",
    "        focal = list(rows.loc[:,f\"class_site_{focal_site}\"])\n",
    "        local = list(rows.loc[:,f\"class_{col}\"])\n",
    "        \n",
    "        # Make a contigency table for these classes between sites 63 and 37\n",
    "        table = np.zeros((2,2),dtype=int)\n",
    "        for i in range(len(rows)):\n",
    "            if focal[i] in focal_classes.keys() and local[i] in local_classes.keys():\n",
    "                table[local_classes[local[i]],focal_classes[focal[i]]] += 1\n",
    "                \n",
    "        # Run fisher's exact test on contingency table\n",
    "        p_value = stats.fisher_exact(table)[1]\n",
    "        out.append(f\"{p_value:.15f}\\n\")\n",
    "\n",
    "        # Create a dictionary mapping this table to human-readable names\n",
    "        coincidence_pie = {}\n",
    "        for x in focal_classes:\n",
    "            for y in local_classes:\n",
    "                out.append(f\"{x} {y} {table[local_classes[y],focal_classes[x]]}\\n\")\n",
    "                coincidence_pie[(x,y)] = table[local_classes[y],focal_classes[x]]\n",
    "        \n",
    "        # Calculate expected number of aro/aro pairs\n",
    "        site_37_aro = 0\n",
    "        site_63_aro = 0\n",
    "        all_seen = 0\n",
    "        for k in coincidence_pie:\n",
    "            if k[0] == \"aromatic\":\n",
    "                site_37_aro += coincidence_pie[k]\n",
    "            if k[1] == \"aromatic\":\n",
    "                site_63_aro += coincidence_pie[k]\n",
    "\n",
    "            all_seen += coincidence_pie[k]\n",
    "\n",
    "        E = all_seen*(site_37_aro/all_seen * site_63_aro/all_seen)\n",
    "        out.append(f\"expected number of aro/aro: {E:.3f}\\n\")\n",
    "    \n",
    "        # Figure out coloring for the coincidence pie chart\n",
    "        colors = []\n",
    "        for k in coincidence_pie.keys():\n",
    "            try:\n",
    "                colors.append(coincidence_colors[k])\n",
    "            except KeyError:\n",
    "                colors.append(\"black\")\n",
    "        \n",
    "        # Plot coincidence pie chart\n",
    "        ax[2,counter].pie(coincidence_pie.values(),labels=coincidence_pie.keys(),colors=colors)\n",
    "        ax[2,counter].set_title(f\"{p_value:.3e}\")\n",
    "        \n",
    "        counter +=1 \n",
    "        \n",
    "    out.append(\"\\n\")\n",
    "        \n",
    "    plt.savefig(f\"{para}.pdf\")\n",
    "\n",
    "print(\"\".join(out))\n",
    "f = open('output-stats.txt','w')\n",
    "f.write(\"\".join(out))\n",
    "f.close()\n",
    "\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a73bebe8-649b-47ac-a176-337c39fbcd8e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c192053c-5f98-4941-b322-46d7cb4f5df1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
